### PR TITLE
Make unStream use rstState, making it safer.

### PR DESCRIPTION
**Pull request by pranaysashank**

### CODE REVIEW

File: src/Streamly/Streams/StreamK.hs [hunk 1]
```diff
@@ -912,7 +908,7 @@ takeWhile p m = fromStream $ go (toStream m)
                        | otherwise = stp
             yieldk a r | p a       = yld a (go r)
                        | otherwise = stp
-         in unStream m1 (rstState st) stp single yieldk
+         in unStream m1 st stp single yieldk
 
 drop :: IsStream t => Int -> t m a -> t m a
 drop n m = fromStream $ Stream $ \st stp sng yld ->
```
Comment (harendra-kumar) [inline src/Streamly/Streams/StreamK.hs]: missed removing the "rstState" below?

Comment (pranaysashank) [reply src/Streamly/Streams/StreamK.hs]: Oh yeah, seemed to have missed it. Should use `unStreamShared` in yieldk instead.

**Comment by harendra-kumar** [comment]
There are some regressions in some multi-stream benchmarks after some commits that I pushed:

```
Benchmark                             default(0)(μs)(base) default(1)(%)(-base)
------------------------------------- -------------------- --------------------
serially/multi-stream/isPrefixOf                    122.49              +122.81
serially/multi-stream/eqBy                          122.48               +43.88
serially/multi-stream/cmpBy                         153.41               +29.91
```

These regression seem to go away by seemingly unrelated changes e.g. by removing commit 90844f9 . They also go away if we remove the pattern synonym at commit 315abcf. Since I could not blame any specific change, I am accepting these.

**Comment by harendra-kumar** [comment]
All the commits on this branch till now, compared with master, only regressions and improvements are shown in the results below, the rest of the benchmarks remain more or less the same:

Linear benchmark - multi-stream/stripPrefix improved while multi-stream/isSubsequenceOf degraded by the same degree:

```
Benchmark                             default(0)(μs)(base) default(1)(%)(-base)
------------------------------------- -------------------- --------------------
serially/multi-stream/isSubsequenceOf               126.52              +107.24
serially/iterated/scanl1(1/100)                     202.04                +5.25
serially/iterated/dropWhileTrue                     172.79                +4.93
serially/pure/id                                     31.71                +4.20
serially/filteringX4/dropWhile-true                 184.86                +4.12
serially/generation/fracFromThenTo                  197.74                +3.50
serially/elimination/foldl1'                        825.66                +3.37
serially/transformationX4/fmap                       61.85                +3.24
serially/pure/eqBy                                   61.98                +3.04
...
serially/multi-stream/zipM                         1008.85                -3.03
serially/pure/cmpBy                                  86.48                -3.13
serially/mixedX4/drop-map                            32.66                -3.54
serially/multi-stream/concatMap                    1520.96                -3.94
serially/generation/fromFoldable                   1359.59                -3.99
serially/pure/foldl'                                 41.96                -4.01
serially/iterated/takeAll                          3472.07                -4.20
serially/elimination/uncons                         876.70                -5.20
serially/mixedX4/filter-scan                         70.66                -5.55
serially/pure/readsPrec                             900.02                -6.60
serially/multi-stream/mergeBy                      1082.14                -6.63
serially/multi-stream/zip                           981.70                -7.61
serially/elimination/foldr1                       13439.28               -23.95
serially/multi-stream/stripPrefix                   278.03               -52.78
serially/elimination/foldr                        31126.81               -66.95
serially/elimination/foldrM                       30989.89               -68.05
```

Linear-async - the `mapM` regressions are not real regressions, that's because of a change in the way we benchmark it, this is because of using concurrent `unfoldrM` for generation of stream:

```
Benchmark                               default(0)(ms)(base) default(1)(%)(-base)
--------------------------------------- -------------------- --------------------
parallely/mapM                                          9.05              +308.29
wAsyncly/mapM                                           3.22               +91.72
aheadly/mapM                                            4.60               +90.10
asyncly/mapM                                            3.07               +89.90

parallely/foldMapWithM                                  8.84                +5.68
aheadly/foldMapWithM                                    4.63                +5.17
parallely/zip                                           5.52                +3.60
...
wAsyncly/fromFoldableM                                  1.94                -3.09
parallely/zipM                                          6.00                -3.98
wAsyncly/foldMapWithM                                   3.21               -32.23
asyncly/foldMapWithM                                    3.21               -33.09
```

nested benchmark - only improvements:

```
Benchmark                default(0)(μs)(base) default(1)(%)(-base)
------------------------ -------------------- --------------------
...
wSerially/filterAllOut                 592.71                -3.23
parallely/filterAllIn                37013.67                -4.88
parallely/breakAfterSome             36774.79                -5.24
serially/filterAllOut                  594.18                -5.31
serially/filterSome                    600.74                -7.24
parallely/toList                     40048.63                -8.14
wSerially/breakAfterSome               906.24                -8.25
serially/breakAfterSome                674.22                -8.52
wSerially/filterAllIn                  901.90                -9.65
parallely/toNull                     38104.18                -9.80
serially/toNull                        680.60               -11.54
wSerially/filterSome                   590.50               -12.34
serially/filterAllIn                   689.48               -12.41
aheadly/filterAllOut                  8366.04               -12.73
wSerially/toNull                       884.39               -13.40
aheadly/filterSome                    8319.18               -15.03
asyncly/filterAllOut                  3999.14               -19.61
asyncly/filterSome                    4245.63               -29.17
wAsyncly/filterSome                   4453.90               -30.22
asyncly/toList                       10824.06               -31.82
wAsyncly/filterAllOut                 4512.37               -32.69
wAsyncly/toList                      11810.43               -34.39
asyncly/toNull                        9691.97               -34.64
wAsyncly/filterAllIn                  9572.47               -35.54
wAsyncly/breakAfterSome               9863.95               -36.40
wAsyncly/toNull                      10363.32               -38.74
asyncly/breakAfterSome               10358.58               -45.54
asyncly/filterAllIn                  10561.32               -45.90
```

**Pull request merged**